Concurrency: Generalize UnblockCallback to MachineCallback

shamb0 opened the pull request:
## Key Changes:
    - Introduces `MachineCallbackState` to represent operation outcomes (Ready/TimedOut)
    - Replaces `UnblockCallback` with a more general `MachineCallback` trait
    - Consolidates callback methods into a single interface for better extensibility
    - Updates thread blocking mechanisms and synchronization primitives to use the new callback system

## This is the first PR in a series to add support for asynchronous file operations:

- **Reference**: Requirements Discussion at https://github.com/rust-lang/miri/pull/4084#discussion_r1894581416

    1. Generalize callback system (this PR)
    2. Add callback support to FileDescription::read
    3. Implement readv using the new callback system

shamb0 commented:
@rustbot ready

RalfJung commented on src/concurrency/thread.rs:
```diff
@@ -38,70 +38,62 @@ pub enum TlsAllocAction {
 }
-pub type DynUnblockCallback<'tcx> = Box<dyn UnblockCallback<'tcx> + 'tcx>;
 
+pub type DyMachineCallback<'tcx> = Box<dyn MachineCallback<'tcx> + 'tcx>;
```
Please call this `DynMachineCallback<'tcx, T>`.

And then define `DynUnblockCallback<'tcx>` as alias for `DynMachineCallback<'tcx, UnblockKind>`.

RalfJung commented on src/concurrency/thread.rs:
```diff
@@ -38,70 +38,62 @@ pub enum TlsAllocAction {
+}
+
+/// Generic callback trait for machine operations.
+pub trait MachineCallback<'tcx>: VisitProvenance {
```
Please move this to `machine.rs`.

RalfJung replied:
```suggestion
pub trait MachineCallback<'tcx, T>: VisitProvenance {
```

RalfJung commented on src/concurrency/thread.rs:
```diff
@@ -38,70 +38,62 @@ pub enum TlsAllocAction {
+    fn call(
+        self: Box<Self>,
+        ecx: &mut InterpCx<'tcx, MiriMachine<'tcx>>,
+        result: MachineCallbackState,
```
The argument type should not be hard-coded.
```suggestion
        arg: T,
```

RalfJung commented on src/concurrency/thread.rs:
```diff
@@ -38,70 +38,62 @@ pub enum TlsAllocAction {
+/// The completion state of an asynchronous machine operation in Miri's concurrency system.
+/// Used by callbacks to determine how a blocked thread should proceed after being woken.
+#[derive(Debug)]
+pub enum MachineCallbackState {
```
```suggestion
pub enum UnblockKind {
```

RalfJung commented on src/concurrency/thread.rs:
```diff
@@ -38,70 +38,62 @@ pub enum TlsAllocAction {
-    /// thread was blocking on having occurred.
-    fn timeout(self: Box<Self>, _ecx: &mut InterpCx<'tcx, MiriMachine<'tcx>>)
-    -> InterpResult<'tcx>;
+/// The completion state of an asynchronous machine operation in Miri's concurrency system.
+/// Used by callbacks to determine how a blocked thread should proceed after being woken.
```
```suggestion
/// The argument type for the "unblock" callback, indicating why the thread got unblocked.
```

RalfJung commented on src/concurrency/thread.rs:
```diff
@@ -38,70 +38,62 @@ pub enum TlsAllocAction {
+pub enum MachineCallbackState {
+    /// The operation completed successfully and the thread can proceed with its normal execution.
+    Ready,
+    /// The operation did not complete within its specified duration and should handle the timeout case.
```
```suggestion
    /// The operation did not complete within its specified duration.
```

RalfJung commented on src/concurrency/thread.rs:
```diff
@@ -38,70 +38,62 @@ pub enum TlsAllocAction {
-        @timeout = |$this_timeout:ident| $timeout:block
-    ) => {{
+    (@capture<$tcx:lifetime $(,)? $($lft:lifetime),*> { $($name:ident: $type:ty),* $(,)? }
+     @unblock = |$this:ident, $result:ident| $complete:expr $(,)?)
```
```suggestion
     @unblock = |$this:ident, $arg:ident| $body:expr $(,)?)
```

RalfJung commented on src/concurrency/thread.rs:
```diff
@@ -38,70 +38,62 @@ pub enum TlsAllocAction {
+
+/// Macro for creating machine callbacks
 #[macro_export]
 macro_rules! callback {
```
This should also be in `machine.rs`

RalfJung commented on src/concurrency/thread.rs:
```diff
@@ -656,11 +648,18 @@ impl<'tcx> ThreadManager<'tcx> {
-                    @unblock = |this| {
-                        if let Some(data_race) = &mut this.machine.data_race {
-                            data_race.thread_joined(&this.machine.threads, joined_thread_id);
+                    @unblock = |this, tcb_state| {
+                        match tcb_state {
```
```suggestion
                    @unblock = |this, unblock| {
```
Please also use this name for all the other unblock callbacks.

RalfJung commented on src/concurrency/thread.rs:
```diff
@@ -656,11 +648,18 @@ impl<'tcx> ThreadManager<'tcx> {
+                                interp_ok(())
+                            }
+                            MachineCallbackState::TimedOut => {
+                                panic!("Join thread operation received unexpected timeout state - operations do not support timeouts")
```
No need to write such a detailed error each time. Also, please avoid the rightwards drift. So instead of a `match unblock`, just have something like
```rust
assert_eq!(unblock, UnblockKind::Ready);
```
at the beginning of the function, and then go on with the function body as before.

RalfJung commented:
Thanks for the PR! I left some comments.

@rustbot author

shamb0 commented:
Thank you for your thoughtful review and detailed comments. I've addressed all the feedback to improve the code's quality, ensuring the changes align with your recommendations on callback generics.  

The PR is now ready for your review, but I’m happy to make further refinements if needed to meet our quality standards.  

Please feel free to reach out if any part of the changes needs clarification or further attention. :)

RalfJung commented on src/concurrency/thread.rs:
```diff
@@ -38,71 +38,20 @@ pub enum TlsAllocAction {
-            }
-        }
+/// Type alias for boxed machine callbacks with generic argument type.
+pub type DyMachineCallback<'tcx, T> = Box<dyn MachineCallback<'tcx, T> + 'tcx>;
```
This type alias should also be in `machine.rs`, next to the definition of `MachineCallback`. Also, please consistently use the `Dyn` prefix, not `Dy`.

RalfJung commented on src/machine.rs:
```diff
@@ -1723,3 +1723,77 @@ impl<'tcx> Machine<'tcx> for MiriMachine<'tcx> {
     }
 }
+
+/// Trait for callbacks handling asynchronous machine operations.
+///
+/// Callbacks receive a completion state and can perform follow-up actions while
+/// maintaining interpreter invariants. They are executed with mutable access to
+/// the interpreter context.
+///
+/// # Type Parameters
+/// - `'tcx`: Typing context lifetime for the interpreter.
+/// - `T`: Type of argument passed to the callback on completion.
```
```suggestion
/// Trait for callbacks handling asynchronous machine operations.
```
There's a lot of text here but it doesn't say much, so IMO it's not worth it.

Btw, are you using LLMs to generate (parts of) your comments? Knowing where you are using LLMs would help me in the review. :) As you know, LLMs are prone to generate plausible-sounding but subtly wrong text, so one has to review them with a different kind of care than human-written comments.

RalfJung commented on src/machine.rs:
```diff
@@ -1723,3 +1723,77 @@ impl<'tcx> Machine<'tcx> for MiriMachine<'tcx> {
+/// - `'tcx`: Typing context lifetime for the interpreter.
+/// - `T`: Type of argument passed to the callback on completion.
+pub trait MachineCallback<'tcx, T>: VisitProvenance {
+    /// Executes the callback when an operation completes.
+    ///
+    /// # Arguments
+    /// - `self`: Owned callback, boxed for dynamic dispatch
+    /// - `ecx`: Mutable interpreter context
+    /// - `arg`: Operation-specific completion argument
+    ///
+    /// # Returns
+    /// Success or error of the callback execution.
```
```suggestion
    /// The function to be invoked when the callback is fired.
```

RalfJung commented on src/machine.rs:
```diff
@@ -1723,3 +1723,77 @@ impl<'tcx> Machine<'tcx> for MiriMachine<'tcx> {
+    ) -> InterpResult<'tcx>;
+}
+
+/// Creates machine callbacks with captured variables and generic argument types.
+///
+/// This macro generates the boilerplate needed to create type-safe callbacks:
+/// - Creates a struct to hold captured variables
+/// - Implements required traits (VisitProvenance, MachineCallback)
+/// - Handles proper lifetime and type parameters
+///
+/// The callback body receives the interpreter context and completion argument.
```
Again this is a lot of text but very helpful text. Instead, can you put a small example of how to use the macro? That would be a lot more useful.

RalfJung commented on src/machine.rs:
```diff
@@ -1723,3 +1723,77 @@ impl<'tcx> Machine<'tcx> for MiriMachine<'tcx> {
+macro_rules! callback {
+    (@capture<$tcx:lifetime $(,)? $($lft:lifetime),*>
+     { $($name:ident: $type:ty),* $(,)? }
+     @unblock = |$this:ident, $arg:ident: $arg_ty:ty| $body:expr $(,)?) => {{
```
Please remove the `@unblock =` everywhere. Now that there's only one function body here, it should not be needed any more.

RalfJung commented on src/machine.rs:
```diff
@@ -1723,3 +1723,77 @@ impl<'tcx> Machine<'tcx> for MiriMachine<'tcx> {
+    (@capture<$tcx:lifetime $(,)? $($lft:lifetime),*>
+     { $($name:ident: $type:ty),* $(,)? }
+     @unblock = |$this:ident, $arg:ident: $arg_ty:ty| $body:expr $(,)?) => {{
+        // Create callback struct with the captured variables and generic type T
```
Please remove these comments, they don't add any information that's not already immediately obvious from looking at the code.

Comments are supposed to describe the code in a high-level way, helping the reader understand the goal of the low-level details visible in the code. These comments feel LLM-generated, and LLMs are pretty bad at generating good comments -- writing good comments requires actually understanding the code, which LLMs are not capable of.

RalfJung commented on src/machine.rs:
```diff
@@ -1723,3 +1723,77 @@ impl<'tcx> Machine<'tcx> for MiriMachine<'tcx> {
+            _phantom: std::marker::PhantomData<&$tcx ()>,
+        }
+
+        // Implement VisitProvenance trait for the callback
```
```suggestion
```

RalfJung commented on src/machine.rs:
```diff
@@ -1723,3 +1723,77 @@ impl<'tcx> Machine<'tcx> for MiriMachine<'tcx> {
+            }
+        }
+
+        // Implement MachineCallback trait with the specified argument type
```
```suggestion
```

RalfJung commented on src/shims/time.rs:
```diff
@@ -331,8 +331,14 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
-                @unblock = |_this| { panic!("sleeping thread unblocked before time is up") }
-                @timeout = |_this| { interp_ok(()) }
+                @unblock = |_this, unblock: UnblockKind| {
+                    match unblock {
```
This could be an `assert_eq!(unblock, UnblockKind::TimedOut);` (same below)

RalfJung reviewed (commented):
Thanks! This is pretty close.

However, I am not entirely convinced by some of the comments you have been adding. Were they generated by an LLM? See the detailed review for more.

Generating comments with an LLM is fine but they should be reviewed for plausibility and usefulness before putting them in the codebase. I helps me do that review if you let me know which parts are LLM-generated.

RalfJung commented:
There are merge conflicts due to other PRs that landed recently in this repo. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

@rustbot author

shamb0 commented:
Thank you, @RalfJung, for your detailed feedback on the documentation. I appreciate your emphasis on keeping explanations concise, high-level, and focused on code intent rather than implementation details.  

I've updated the documentation based on your suggestions and will continue applying these principles to future Miri contributions. While I used an LLM to assist with refining the docstrings, I ensured the revisions align with your guidance to avoid verbosity and redundancy.  

Your earlier review comments have been addressed, but please let me know if there’s anything else that needs adjustment. I’m also working on gaining a deeper understanding of Miri’s documentation practices to ensure consistency across the codebase.

RalfJung commented:
This looks great, thanks!

@bors r+

shamb0 commented:
> This looks great, thanks! I just added some doc tweaks.

Thankyou
